src/psi/j/executors/saga.py

```diff
@@ -0,0 +1,222 @@
+    rs.job.FAILED: JobState.FAILED,
+    rs.job.CANCELED: JobState.CANCELED,
+}
+
+
+class _JobMapping:
+    def __init__(self, jpsi_job: Job, saga_job: Optional[rs.job.Job] = None):
+        self.jpsi_job = jpsi_job
+        self.saga_job = saga_job
+
+
+class SagaExecutor(JobExecutor):
+    """
+    An executor that use SAGA to run jobs.
+
+    For details about SAGA, see https://en.wikipedia.org/wiki/Simple_API_for_Grid_Applications
+    and related links.
+    """
+
+    assert('job' in dir(rs)), 'incomplete installation'
+
+    def __init__(self, url: Optional[str] = None,
+                 config: Optional[JobExecutorConfig] = None) -> None:
+        """
+        Initializes a `SagaExecutor`.
+
+        Parameters
+        ----------
+        url
+            An URL representing a SAGA resource manager to connect to.
+        config
+            Not used.
+        """
+        super().__init__(url=url, config=config)
+
+        self._jobs = dict()  # type: Dict[str, _JobMapping]
+        self._lock = RLock()
+        self._js = rs.job.Service(self.url)
+        self._cb = None
+        atexit.register(self._close)
+
+    def _close(self) -> None:
+        if self._js:
+            self._js.close()
+
+    def __del__(self) -> None:
+        """
+        Removes process shutdown hooks added earlier.
+
+        The hooks are used to close the SAGA service used by this executor.
+        """
+        atexit.unregister(self._close)
+        self._close()
+
+    def _state_cb(self, saga_job: rs.job.Job, metric: object, ctx: object) -> bool:
+        jpsi_uid = saga_job._jex_uid
+        jpsi_job = self._get_jpsi_job(jpsi_uid)
+        ec = None
+
+        if saga_job.state in [rs.FAILED, rs.DONE]:
+            ec = saga_job.exit_code
+
+        logger.debug('%s --> %s', jpsi_uid, saga_job.state)
+
+        job_status = JobStatus(_STATE_MAP[saga_job.state], time.time(), exit_code=ec)
+        jpsi_job._set_status(job_status, self)
+        if self._cb:
+            self._cb.job_status_changed(jpsi_job, job_status)
+
+        return True
+
+    def submit(self, job: Job) -> None:
+        """
+        Submits the specified :class:`~psi.j.Job` to be run using SAGA.
+
+        Parameters
+        ----------
+        job
+            The :class:`~psi.j.Job` to be submitted.
+        """
+        # derive SAGA job description and submit it
+        job_mapping = _JobMapping(job)
+        with self._lock:
+            self._jobs[job.id] = job_mapping
+        job._executor = self
+
+        jd = self._job_2_descr(job)
+        saga_job = self._js().create_job(jd)
+        saga_job._jex_uid = job.id
+
+        job_mapping.saga_job = saga_job
+        saga_job.add_callback(rs.STATE, self._state_cb)
+
+        # TODO: the backend attribute is not standard; why do we need it?
+        self._update_job_status(job, JobStatus(JobState.QUEUED, metadata={'backend': self.url}))
+
+        self._update_job_status(job, JobStatus(JobState.ACTIVE))
```

Comment:
I'm curious as to why the job immediately gets moved to `ACTIVE`.  Shouldn't that be handled by the state_cb since the job could be stuck in the queue?  Maybe this is just a behavior of SAGA that I'm unaware of.

Reply:
I agree with that observation - the SAGA's `RUNNING` state should translate to `ACTIVE` (as also is encoded in the `_STATE_MAP`).